Software/EmonESP/src_6chan/mqtt.cpp

```diff
@@ -86,50 +97,51 @@ boolean mqtt_connect()
 // base topic = emon/emonesp
 // MQTT Publish: emon/emonesp/CT1 > 3935 etc..
 // -------------------------------------------------------------------
-void mqtt_publish(String data)
+void mqtt_publish(const char * data)
 {
-  String mqtt_data = "";
-  String topic = mqtt_topic + "/" + mqtt_feed_prefix;
-  int i = 0;
-  while (int (data[i]) != 0)
+  const char * data_ptr = data;
+  char * topic_ptr = mqtt_topic_prefix;
+  topic_ptr += sprintf(mqtt_topic_prefix, "%s/%s", mqtt_topic, mqtt_feed_prefix);
+
+  do
   {
-    // Construct MQTT topic e.g. <base_topic>/CT1 e.g. emonesp/CT1
-    while (data[i] != ':') {
-      topic += data[i];
-      i++;
-      if (int(data[i]) == 0) {
-        break;
-      }
-    }
-    i++;
-    // Construct data string to publish to above topic
-    while (data[i] != ',') {
-      mqtt_data += data[i];
-      i++;
-      if (int(data[i]) == 0) {
-        break;
-      }
+    int pos = strcspn(data_ptr, ":");
+    strncpy(topic_ptr, data_ptr, pos);
+    topic_ptr[pos] = 0;
+    data_ptr += pos;
+    if (*data_ptr++ == 0) {
+      break;
     }
+
+    pos = strcspn(data_ptr, ",");
+    strncpy(mqtt_data, data_ptr, pos);
+    mqtt_data[pos] = 0;
+    data_ptr += pos;
+
     // send data via mqtt
     //delay(100);
-    //DBUGS.printf("%s = %s\r\n", topic.c_str(), mqtt_data.c_str());
-    mqttclient.publish(topic.c_str(), mqtt_data.c_str());
-    topic = mqtt_topic + "/" + mqtt_feed_prefix;
-    mqtt_data = "";
-    i++;
-    if (int(data[i]) == 0) break;
-  }
+    //DBUGS.printf("%s = %s\r\n", mqtt_topic_prefix, mqtt_data);
+    if (!mqttclient.publish(mqtt_topic_prefix, mqtt_data))
+    {
+      return;
+    }
+  } while (*data_ptr++ != 0);
 
   // send esp free ram
-  String ram_topic = mqtt_topic + "/" + mqtt_feed_prefix + "freeram";
-  String free_ram = String(ESP.getFreeHeap());
-  mqttclient.publish(ram_topic.c_str(), free_ram.c_str());
+  sprintf(mqtt_topic_prefix, "%s/%sfreeram", mqtt_topic, mqtt_feed_prefix);
+  sprintf(mqtt_data, "%lu", ESP.getFreeHeap());
+  if (!mqttclient.publish(mqtt_topic_prefix, mqtt_data))
+  {
+    return;
+  }
 
   // send wifi signal strength
-  long rssi = WiFi.RSSI();
-  String rssi_S = String(rssi);
-  String rssi_topic = mqtt_topic + "/" + mqtt_feed_prefix + "rssi";
-  mqttclient.publish(rssi_topic.c_str(), rssi_S.c_str());
+  sprintf(mqtt_topic_prefix, "%s/%srssi", mqtt_topic, mqtt_feed_prefix);
+  sprintf(mqtt_data, "%d", WiFi.RSSI());
+  if(!mqttclient.publish(mqtt_topic_prefix, mqtt_data))
+  }
```

Comment:
This should be a {

Reply:
Yes, it should.  I incorrectly moved this change from my other branch.